nats.go

```diff
@@ -147,6 +147,9 @@ type ErrHandler func(*Conn, *Subscription, error)
 // return the base64 encoded signature.
 type SignatureHandler func([]byte) []byte
 
+// TokenHandler is used to generate a new token for authentication.
+type TokenHandler func() string
```

Comment:
I added @kozlovic to that one since he maintains the C client.

Reply:
I would advise against that. The token callback is invoked synchronously from the connect/reconnect thread under the connection lock, which means that if user invokes almost any connection's function, then it will deadlock (since there is no re-entrant lock support).

It is actually something that should be clearly stated for users planning to use this callback.

Alternatively, we would have to release/reacquire lock around the invocation of the callback, but that scares me since there is state that then could change and may need to be re-evaluated upon reacquiring the lock.

Reply:
Thanks a lot. Very good point!